commit 7d2cf1c0c9df0d126fafaa50c13138781b78c98b
Author: Damien Le Moal <damien.lemoal@wdc.com>
Date:   Thu Jan 12 15:25:10 2017 +0900

    scsi: sd: Fix wrong DPOFUA disable in sd_read_cache_type
    
    
    [ Upstream commit 26f2819772af891dee2843e1f8662c58e5129d5f ]
    
    Zoned block devices force the use of READ/WRITE(16) commands by setting
    sdkp->use_16_for_rw and clearing sdkp->use_10_for_rw. This result in
    DPOFUA always being disabled for these drives as the assumed use of
    the deprecated READ/WRITE(6) commands only looks at sdkp->use_10_for_rw.
    Strenghten the test by also checking that sdkp->use_16_for_rw is false.
    
    Signed-off-by: Damien Le Moal <damien.lemoal@wdc.com>
    Reviewed-by: Hannes Reinecke <hare@suse.com>
    Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>
    Signed-off-by: Sasha Levin <alexander.levin@verizon.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

diff --git a/drivers/scsi/sd.c b/drivers/scsi/sd.c
index 931af07..13ac7e5 100644
--- a/drivers/scsi/sd.c
+++ b/drivers/scsi/sd.c
@@ -2572,7 +2572,8 @@ sd_read_cache_type(struct scsi_disk *sdkp, unsigned char *buffer)
 		if (sdp->broken_fua) {
 			sd_first_printk(KERN_NOTICE, sdkp, "Disabling FUA\n");
 			sdkp->DPOFUA = 0;
-		} else if (sdkp->DPOFUA && !sdkp->device->use_10_for_rw) {
+		} else if (sdkp->DPOFUA && !sdkp->device->use_10_for_rw &&
+			   !sdkp->device->use_16_for_rw) {
 			sd_first_printk(KERN_NOTICE, sdkp,
 				  "Uses READ/WRITE(6), disabling FUA\n");
 			sdkp->DPOFUA = 0;
